### DISCUSSION TITLE
tud_task frequency

### DISCUSSION THREAD

**Discussion by slc-tl** [hathach/tinyusb#2028] [Q&A] [answered]
Hi Everyone, 

First of all I want to say how impressed I've been with the tinyusb library. It's super-useful -- thank you to all the contributors! 

On to my question:

I've noticed in the cdc_msc_freertos that the `usb_device_task` is free-running. Unless I'm mistaken, there is no synchronization, semaphore, or delay that would let the scheduler give the CPU to another task than the `usb_device_task`. The `usb_device_task` also holds the highest priority of the two freeRTOS tasks in the cdc_msc_freertos example -- so it's unlikely to be preempted as well. 

I'm working on a project that uses USB device communication alongside other interfaces on a fairly performant processor (ARM Cortex M7, 300 MHz). Ideally this processor would be able to manage other activities than the USB stack. With that in mind, I have the following questions: 

- What frequency does `tud_task()` need to be called for the tinyusb device stack to operate correctly? 
- Does the frequency with which `tud_task` needs to be called change depending on the state of the USB? It seems like it needs to be called very frequently during the SETUP phase, but I imagine that during other states `tud_task()` can be called much less frequently. Perhaps a table in the documentation outlining these required frequencies exists? 
- Is there any capability shown in any of the examples to signify that `tud_task` needs to be called through an interrupt that is deferred to the RTOS rather than polling `tud_task`? I suppose I could piggyback the USB IRQ but I'm curious if a more abstract intervention exists. 

Thanks!

**Comment by DRNadler** [comment]
tud_task blocks on entry waiting for events, and only returns when it has processed one or more events. Thus the CPU is normally doing something else or waiting. Normally tud_task is called in a task that loops forever (calling tud_task then processing any interesting side-effects **_without taking too long to process side-effects_**). As long as this is a high-priority task, you shouldn't have to be concerned about the actual frequency, no? I'm not exactly answering your question regarding frequency, but hope this helps!

**Reply by slc-tl** [reply]
Is the blocking mechanism significant to the freeRTOS scheduler? I.e. will the freeRTOS scheduler recognize that this task is blocked so that the CPU is yielded?

**Reply by DRNadler** [reply, marked as answer]
osal_queue_receive should be implemented this way for all supported RTOS; see for example osal_freertos.h.
Note that osal_queue_receive is called with an 'infinite' timeout so should not return unless event (or error).
Hope that's clear!
PS: Feel free to make a pull-request documenting this in the headers and/or documentation so the next guy doesn't have to dig...

**Reply by slc-tl** [reply]
Yup that's right, you got to it faster than I did!

For others reading this, you need to set `#define CFG_TUSB_OS  OPT_OS_FREERTOS` in tusb_config.h to ensure that the osal_freertos.h is the osal header pointed to during compilation. 

Appreciate the help here, I'll absolutely try to follow up on the documentation PR. Do you have any recommendation of where a detail like this should go in the documentation?

**Reply by DRNadler** [reply]
Getting Started documentation, osal.h header, usbd.c comments clarifying loop??

**Comment by DRNadler** [comment]
PS: Note this issue is a repeat (thus indicating a documentation need): #1972
